### DISCUSSION TITLE
OpenSSL 1.1.1q failing to build

### DISCUSSION THREAD

**Discussion by stack** [rbenv/ruby-build#2009] [Build failures] [answered]
This is a fresh install of macOS 12.4 on a M1 Pro MacBook Pro. Attempting to install Ruby 3.1.2.

* Xcode 13.4.1
* Homebrew installed as of today (2022-07-11).
* ruby-build 20220710

The OpenSSL 1.1.1q fails to build in the tests:

```
Downloading openssl-1.1.1q.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/d7939ce614029cdff0b6c20f0e2e5703158a489a72b2507b8bd51bf8c8fd10ca
Installing openssl-1.1.1q...

BUILD FAILED (macOS 12.4 using ruby-build 20220710)

Inspect or clean up the working tree at /var/folders/4y/m2qk7qcx099b6g_zt98mp3jh0000gn/T/ruby-build.20220711120126.51360.08iZ3t
Results logged to /var/folders/4y/m2qk7qcx099b6g_zt98mp3jh0000gn/T/ruby-build.20220711120126.51360.log

Last 10 log lines:
clang  -Iinclude -arch arm64 -O3 -Wall -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/Users/stephen/.rbenv/versions/3.1.2/include  -MMD -MF test/v3ext.d.tmp -MT test/v3ext.o -c -o test/v3ext.o test/v3ext.c
clang  -Iinclude -arch arm64 -O3 -Wall -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/Users/stephen/.rbenv/versions/3.1.2/include  -MMD -MF test/v3nametest.d.tmp -MT test/v3nametest.o -c -o test/v3nametest.o test/v3nametest.c
test/v3ext.c:201:24: error: implicitly declaring library function 'memcmp' with type 'int (const void *, const void *, unsigned long)' [-Werror,-Wimplicit-function-declaration]
        if (!TEST_true(memcmp(ip1->data, ip2->data, ip1->length) <= 0))
                       ^
test/v3ext.c:201:24: note: include the header <string.h> or explicitly provide a declaration for 'memcmp'
1 error generated.
make[1]: *** [test/v3ext.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

**Comment by eregon** [comment]
It's a bug in openssl upstream: https://github.com/openssl/openssl/issues/18720

**Comment by eregon** [comment, marked as answer]
`OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration ruby-build/rbenv/asdf ...` helps and works around the missing include in openssl.

The default of `-Werror,-Wimplicit-function-declaration` is macOS-specific (so this does not happen on Linux).

An easy fix for 3.1.2 would be to install openssl 3 instead of openssl 1.1, thoughts on that @mislav? => https://github.com/rbenv/ruby-build/pull/2000

**Reply by abhishekbedi1432** [reply]
Run your ruby install with the argument
`OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration rbenv install [version]`

Reference: https://github.com/rbenv/ruby-build/discussions/2009#discussioncomment-3141255

**Reply by jetdotrb** [reply]
Thanks! It works for me!

**Reply by eregon** [reply]
With the latest release of ruby-build there is no need for this workaround anymore: https://github.com/rbenv/ruby-build/pull/2011

**Reply by dtsdwarak** [reply]
Thanks much for the fix!

**Comment by stack** [comment]
The `OPENSSL_CFLAGS` flags do fix the issue for me. 3.1.2 installed without issue.

**Comment by eregon** [comment]
Alright, then I think we can close this.
Hopefully openssl makes a new 1.1 release soon with the fix so there is no need to work around it.

**Comment by ghost** [comment]
> Could you try if `OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration` helps?
> 
> The default of `-Werror,-Wimplicit-function-declaration` is macOS-specific (so this does not happen on Linux).
> 
> An easy fix for 3.1.2 would be to install openssl 3 instead of openssl 1.1, thoughts on that @mislav? => #2000

i'm sorry. What does this mean. How do I do that?

**Comment by jtethras** [comment]
> > Could you try if `OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration` helps?
> > The default of `-Werror,-Wimplicit-function-declaration` is macOS-specific (so this does not happen on Linux).
> > An easy fix for 3.1.2 would be to install openssl 3 instead of openssl 1.1, thoughts on that @mislav? => #2000
> 
> i'm sorry. What does this mean. How do I do that?

Run your ruby install with the argument
`OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration rbenv install [version]`

**Reply by besi** [reply]
No love here unfortunately. I did open a new issue #2060

**Comment by ghost** [comment]
> OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration

lifesaver. thanks

**Comment by stack** [comment]
20220713 also fixed this issue for Ruby 3.1.x by moving to OpenSSL 3.

**Comment by eregon** [comment]
@mislav @hsbt Do you think we should automatically add the include/patch openssl 1.1.1q before building it?
I wish we would not need to, but openssl doesn't seem to have a made a new release after this broken release and we're already 9 days after the release.
Given it's a trivial and safe change, not related to security I'm starting to consider it, otherwise I would not.

Another thought was could we just revert to 1.1.1p, but that would mean missing a CVE fix (https://www.openssl.org/news/openssl-1.1.1-notes.html), which seems bad.

**Reply by hsbt** [reply]
Can we add `OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration` option to 1.1.1q temporally? I agreed to not use this option by default and don't use 1.1.1p.

**Reply by eregon** [reply]
Right, that's probably a bit easier than to change a file. It's a bit less safe (there could be another missing header on some uncommon OS and that could be problematic, here it's just in tests which are not run so doesn't matter), but I think it's still OK for 1.1.1q.
I guess most operating systems use a patch to deal with this situation, so changing the file is closer to that, but it's not very important.

**Reply by mislav** [reply]
I believe it should be safe to auto-patch `1.1.1q` as long as the patch is taken verbatim from OpenSSL upstream. I think the patch will be a more robust solution than setting environment variables.

**Reply by hsbt** [reply]
We should apply this https://github.com/openssl/openssl/commit/f9e578e720bb35228948564192adbe3bc503d5fb.diff

**Reply by eregon** [reply]
Done in https://github.com/rbenv/ruby-build/pull/2011 (thanks @hsbt!) and released as v20220721

**Comment by aishtiaq7** [comment]
Still having issues even after trying this command: 
`OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration rbenv install 2.7.6` 

![Screenshot 2022-11-23 at 3 15 04 AM](https://user-images.githubusercontent.com/41769418/203533290-35c311a9-2426-4a67-b552-1b2a11661bf4.png)


I'm trying to install React-Native on my MacOS M1, React-Native is requiring 2.7.6 version of ruby. But still seems to run onto this issue. Please help.

**Reply by mislav** [reply]
Hi, without the full log, we cannot tell what happened. Also, note that your build failure doesn't belong in this discussion thread as we were discussing build failures in `1.1.1q`, which had a known bug, and you're building a more recent `1.1.1s` version.

My general advice to everyone building Ruby on macOS would be to install OpenSSL via Homebrew (`brew install openssl@1.1`) and then link Ruby to that:
```sh
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 2.7.6
```

**Reply by hsbt** [reply]
`/usr/local/bin/clang` is not official toolchanins for Apple Silicon. Can you uninstall them and reinstall Ruby 2.7 with rbenv?
